src/osd/OSD.cc

```diff
@@ -10206,43 +10206,80 @@ void OSD::maybe_override_max_osd_capacity_for_qos()
   // osd capacity with the value obtained from running the
   // osd bench test. This is later used to setup mclock.
   if (cct->_conf.get_val<std::string>("osd_op_queue") == "mclock_scheduler") {
-    // Write 200 4MiB objects with blocksize 4KiB
+    int ret;
```

Comment:
i don't quite like the design of

1. calibrate myself on boot up
2. update my setting on monitor

on the contrary, i think this should be driven by external tools like cephadm.

Reply:
> on the contrary, i think this should be driven by external tools like cephadm.

I imagine this to be the eventual goal. However, I will discuss this further with
@neha-ojha and @jdurgin to get more clarity on the way forward. As it stands,
this approach is what we have discussed so far.

Reply:
I don't see external tooling doing this as a goal.

The point of putting it in the OSD is to centralize the logic of handling this, rather than re-implementing it in N different deployment tools. Storing it in the config db makes it easily visible to users, so they can easily see if the performance is not as expected across OSDs.

This is not just a deployment issue, but one of ongoing maintainence as well - disk and machine performance may vary over time. Keeping this in the OSD means it's simple to re-measure it, and since calibration is cheap, it's not an issue for the OSD to decide to do it sometimes during boot.

This design is compatible with external tools doing measurements if we wanted - just have the tool set the setting, and the OSD won't run the benchmark. I'm not seeing a lot of benefits to that approach though. Do you have particular advantages in mind @tchaikov ?

Reply:
no. i don't have any particular advantages in mind. i just like the principal which encodes the machinery at a lower level, and keep the policy or the driver at a higher level so that the result is more likely to be more flexible and elegant.